bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java

```diff
@@ -138,89 +162,147 @@ private synchronized void poll() {
             pollScenes();
         } catch (JsonParseException e) {
             logger.warn("Bridge returned a bad JSON response: {}", e.getMessage());
-        } catch (ProcessingException | IOException e) {
+        } catch (ProcessingException e) {
             logger.warn("Error connecting to bridge: {}", e.getMessage());
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, e.getMessage());
+        } catch (HubMaintenanceException e) {
+            // exceptions are logged in HDPowerViewWebTargets
         }
     }
 
-    private void pollShades() throws JsonParseException, ProcessingException, IOException {
+    private void pollShades() throws JsonParseException, ProcessingException, HubMaintenanceException {
+        HDPowerViewWebTargets webTargets = this.webTargets;
+        if (webTargets == null) {
+            throw new ProcessingException("Web targets not initialized");
+        }
+
         Shades shades = webTargets.getShades();
+        if (shades == null) {
+            throw new JsonParseException("Missing 'shades' element");
+        }
+
+        List<ShadeData> shadesData = shades.shadeData;
+        if (shadesData == null) {
+            throw new JsonParseException("Missing 'shades.shadeData' element");
+        }
+
         updateStatus(ThingStatus.ONLINE);
-        if (shades != null) {
-            Map<String, Thing> things = getThingsByShadeId();
-            logger.debug("Found {} shades", things.size());
-            for (Shade shade : shades.shadeData) {
-                Thing thing = things.get(shade.id);
-                if (thing != null) {
-                    HDPowerViewShadeHandler handler = ((HDPowerViewShadeHandler) thing.getHandler());
-                    if (handler != null) {
-                        logger.debug("Handling update for shade {}", shade.id);
-                        handler.onReceiveUpdate(shade);
-                    } else {
-                        logger.debug("Skipping shade with no handler {}", shade.id);
-                    }
-                } else {
-                    logger.debug("Skipping non-bound shade {}", shade.id);
-                }
-            }
+        logger.debug("Received data for {} shades", shadesData.size());
+
+        Map<String, ShadeData> idShadeDataMap = getIdShadeDataMap(shadesData);
+        Map<Thing, String> thingIdMap = getThingIdMap();
+        for (Entry<Thing, String> item : thingIdMap.entrySet()) {
+            @SuppressWarnings("null")
+            Thing thing = item.getKey();
+            @SuppressWarnings("null")
+            String shadeId = item.getValue();
+            @Nullable
```

Comment:
This is not needed, as the compiler is intelligent enough with local variables.

Reply:
If I remove the `@Nullable` then the Eclipse IDE does give a warning, and then QuickFix suggests the solution "Add @SuppressWarnings 'null' to xyz" analog to above..

Reply:
Strange, I don't get a warning. You could declare the Map's values as Nullable if the Map can contain null values:
```
Map<Thing, @Nullable String> ret = new HashMap<>();
```

Reply:
Done, as above..